proctor-common/src/main/java/com/indeed/proctor/common/ForceGroupsOptionsStrings.java

```diff
@@ -38,39 +47,156 @@ public static ForceGroupsOptions parseForceGroupsString(@Nullable final String f
-            if (piece.charAt(bucketValueStart) != '-') {
-                bucketValueStart++;
+        }
+        return builder.build();
+    }
+
+    private static Optional<Integer> getBucketValueStart(final String groupString) {
+        int bucketValueStart = groupString.length() - 1;
+        for (; bucketValueStart >= 0; bucketValueStart--) {
+            if (!Character.isDigit(groupString.charAt(bucketValueStart))) {
+                break;
             }
-            //  bucketValueStart should now be the index of the minus sign or the first digit in a run of digits going to the end of the word
-            final String testName = piece.substring(0, bucketValueStart).trim();
-            final String bucketValueStr = piece.substring(bucketValueStart);
+        }
+        //  if no name or no value was found, it's not a valid proctor test bucket name
+        if ((bucketValueStart == groupString.length() - 1) || (bucketValueStart < 1)) {
+            return Optional.empty();
+        }
+        //  minus sign can only be at the beginning of a run
+        if (groupString.charAt(bucketValueStart) != '-') {
+            bucketValueStart++;
+        }
+
+        return Optional.of(bucketValueStart);
+    }
+
+    /**
+     * The format of the payloadString is the following:
+     * <ul>
+     *     <li>payloadType:payloadValue</li>
+     *     <li>Where payloadType is one of the 7 payload types supported by Proctor (stringValue, stringArray, doubleValue, longValue, longArray, map)</li>
+     *     <li>If payloadValue is an array is expected in the following format: [value value value]</li>
+     *     <li>If payloadValue is a map expecting following format: [key:value key:value key:value]</li>
+     * </ul>
+     */
+    @Nullable
+    public static Payload parseForcePayloadString(final String payloadString )
+    {
+        final Payload payload = new Payload();
+
+        final String[] splitPayloads = payloadString.split(";");
+        for (final String splitPayload : splitPayloads) {
+            final String[] payloadPieces = payloadString.split(":",2);
             try {
-                final Integer bucketValue = Integer.valueOf(bucketValueStr);
-                builder.putForceGroup(testName, bucketValue);
-            } catch (final NumberFormatException e) {
-                LOGGER.error("Unable to parse bucket value " + bucketValueStr + " as integer", e);
+                final PayloadType payloadType = PayloadType.payloadTypeForName(payloadPieces[0]);
+                String payloadValue = payloadPieces[1];
+                switch (payloadType) {
+                    case DOUBLE_VALUE:
+                    {
+                        payload.setDoubleValue(Double.parseDouble(payloadValue));
+                        break;
+                    }
+                    case DOUBLE_ARRAY:
+                    {
+                        payloadValue = payloadValue.replace("[","").replace("]","");
```

Comment:
this code is repeating for double, long and string array

Reply:
Yes, I moved it into the case statements due to the map being able to have arrays as entries.

Reply:
i mean, can we refactor like this
private String[] getPayloadArray(final String payloadValue) {
  final String payloadValueExtracted = payloadValue.replace("[","").replace("]","");
  return payloadValueExtracted.split(" ");
}

Then in the case statement we can do
payload.setDoubleArray(Arrays.stream(getPayloadArray(payloadValue))
                                .map(Double::valueOf)
                                .toArray(Double[]::new) );

Reply:
Done